fix: essentials.home.bed & essentials.sethome.bed permissions

Hunh0w opened the pull request:
<!--

EssentialsX bug fix submission guide
====================================

NOTE: Failure to fill out this template properly may result in your PR being
      delayed or ignored without warning.

NOTE: Don't type between any arrows in the template, as this text will be
      hidden. This includes this header block and any other explanation text
      blocks.

Want to discuss your PR before submitting it? Join the EssentialsX Development
server: https://discord.gg/CUN7qVb


EssentialsX is GPL
------------------

By contributing to EssentialsX, you agree to license your code under the
GNU General Public License version 3, which can be found at the link below:
https://github.com/EssentialsX/Essentials/blob/2.x/LICENSE


Instructions
------------

If you are submitting a bug fix, please follow the following steps:

1.  Fill out the template in full.
      This includes providing screenshots and a link to the original bug 
      report. If there isn't an existing bug report, we recommend opening a new
      detailed bug report BEFORE opening your PR to fix it, else your PR may be
      delayed or rejected without warning.
      
      You can open a new bug report by following this link:
      https://github.com/EssentialsX/Essentials/issues/new/choose 

2.  When linking logs or config files, do not attach them to the post!
      Copy and paste any logs into https://gist.github.com/, then paste a
      link to them in the relevant parts of the template. Do not use Hastebin
      or Pastebin, as this can cause issues with future reviews.
      
      DO NOT drag logs directly into this text box, as we cannot read these!

3.  If you are fixing a performance issue, please include a link to a
    Timings and/or profiler report, both before and after your PR.

4.  If you are fixing a visual bug, such as in commands, please include
    screenshots so that we can more easily review the proposed fix.
    (You can drag screenshots into the bottom of the editor.)

-->

### Information

<!--
    Replace #nnnn with the number of the original issue. If this PR fixes
    multiple issues, you should repeat the phrase "fixes #nnnn" for each issue. 
-->

This PR fixes #5963. 

### Details

**Proposed fix:**   
If the player uses "/home bed" when he doesn't have the "essentials.home.bed" permission, he's getting a no permission error for it 

**Environments tested:**    

<!-- Type the OS you have used below. -->
OS: Ubuntu 24.04.1 LTS

<!-- Type the JDK version (from java -version) you have used below. -->
Java version: 21

<!--
    Put an "x" inside the boxes for the server software you have tested this 
    bug fix on. If this feature does not apply to a server, strike through the server software using ~~strikethrough~~. If you have tested on other
    environments, add a new line with relevant details.
-->
- [x] Most recent Paper version (1.XX.Y, git-Paper-BUILD)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8


**Demonstration:**    
<!--
    Below this block, include screenshots/log snippets from before and after as
    necessary. If you have created or used a test case plugin, please link to a
    download of the plugin, source code and exact version used where possible.
-->

Hunh0w committed:
fix: essentials.home.bed permission

pop4959 commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandhome.java:
```diff
@@ -73,10 +73,14 @@ public void run(final Server server, final User user, final String commandLabel,
-                        server.getPluginManager().callEvent(event);
-                        if (!event.isCancelled()) {
-                            user.getAsyncTeleport().respawn(charge, TeleportCause.COMMAND, getNewExceptionFuture(user.getSource(), commandLabel));
+                        if(bed == null || finalPlayer.isAuthorized("essentials.home.bed")){
```
You should not be checking if the bed is null here, it is already checked below and has a defined behavior which this change appears to otherwise break.

pop4959 commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandhome.java:
```diff
@@ -73,10 +73,14 @@ public void run(final Server server, final User user, final String commandLabel,
+                            if (!event.isCancelled()) {
+                                user.getAsyncTeleport().respawn(charge, TeleportCause.COMMAND, getNewExceptionFuture(user.getSource(), commandLabel));
+                            }
+                        }else {
```
Please fix the indentation here and in other places in the code to match the code style

pop4959 reviewed (dismissed):
Hey, thanks for making your first PR to Essentials! We appreciate it.

I left a few comments for you to look at. Please also share proof of testing (show before/after screenshot is probably the easiest) in the demonstration section which you left blank.

Also, in the original issue, `essentials.sethome.bed` is mentioned. Have you checked if there is a problem with /sethome or is this working as intended?

Hunh0w committed:
fix: identation & no null bed check

Hunh0w committed:
fix: essentials.sethome.bed permission

Hunh0w committed:
remove: test message

Hunh0w commented:
Hey, thanks you too for your feedback ! ^^
I updated this PR according to your comments.

I added a fix for the `essentials.sethome.bed` there was a problem, we couldn't prevent player from setting his bed position even with this permission to false

Screenshots of the before behaviour:
![Capture d’écran du 2024-12-12 09-03-19](https://github.com/user-attachments/assets/a3658000-d4ee-4c57-9064-4d49da78dbbf)
![Capture d’écran du 2024-12-12 10-11-26](https://github.com/user-attachments/assets/59a92055-bf9a-406e-96cc-b335c141ac64)

Screenshots of the after behaviour:
![Capture d’écran du 2024-12-12 10-09-24](https://github.com/user-attachments/assets/3a70c353-db41-43cf-a5bd-9f853e8a34b2)
![no_bed_perm_result](https://github.com/user-attachments/assets/03ef9313-d803-4f17-a86e-fb556969f444)

For the bed sethome, if we don't have the permission it doesn't produce anything, maybe i should add an error message in this case ?

JRoy committed:
Merge branch '2.x' into 2.x

JRoy commented:
thanks!

Potothingi commented:
Since this commit I can't sleep at night and can't set respawn point. Previous build (2.21.1-dev+18-3b0c229) was fine.
```
essentials.home: true
essentials.home.bed: true
```

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java:
```diff
@@ -823,12 +823,15 @@ public void onPlayerInteract(final PlayerInteractEvent event) {
                             player.sendTl("bedSet", player.getLocation().getWorld().getName(), player.getLocation().getBlockX(), player.getLocation().getBlockY(), player.getLocation().getBlockZ());
                         }
+                    } else if (!isAuthorized) {
+                        event.setCancelled(true);
```
This prevents people interacting with the bed. Why was this changed?

Hunh0w replied:
This was because if someone interact with a bed, they automatically set the bed home even if they don't have the permission
